daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/DataValue.scala

```diff
@@ -0,0 +1,209 @@
+  type DataValueDate = DataValue[DFDLDate, NonNullable with DataValuePrimitiveType]
+  type DataValueTime = DataValue[DFDLTime, NonNullable with DataValuePrimitiveType]
+  type DataValueByteArray = DataValue[Array[Byte], NonNullable with DataValuePrimitiveType]
+  type DataValueBool = DataValue[JBoolean, NonNullable with DataValuePrimitiveType]
+  type DataValueNumber = DataValue[JNumber, NonNullable with DataValuePrimitiveType]
+  type DataValueLong = DataValue[JLong, NonNullable with DataValuePrimitiveType]
+  type DataValueDouble = DataValue[JDouble, NonNullable with DataValuePrimitiveType]
+  type DataValueBigInt = DataValue[JBigInt, NonNullable with DataValuePrimitiveType]
+  type DataValueString = DataValue[JString, NonNullable with DataValuePrimitiveType]
+  type DataValueURI = DataValue[URI, NonNullable with DataValuePrimitiveType]
+  type DataValueFloat = DataValue[JFloat, NonNullable with DataValuePrimitiveType]
+  type DataValueByte = DataValue[JByte, NonNullable with DataValuePrimitiveType]
+  type DataValueInt = DataValue[JInt, NonNullable with DataValuePrimitiveType]
+  type DataValueShort = DataValue[JShort, NonNullable with DataValuePrimitiveType]
+  type DataValueDINode = DataValue[DINode, NonNullable with DataValuePrimitiveType]
+  type DataValueUseNilForDefault = DataValue[UseNilForDefaultObj, NonNullable]
+
+  import scala.language.implicitConversions
+
+  @inline implicit def toDataValue(v: JBigDecimal): DataValueBigDecimal = new DataValue(v)
+  @inline implicit def toDataValue(v: DFDLCalendar): DataValueCalendar = new DataValue(v)
+  @inline implicit def toDataValue(v: DFDLDateTime): DataValueDateTime = new DataValue(v)
+  @inline implicit def toDataValue(v: DFDLDate): DataValueDate = new DataValue(v)
+  @inline implicit def toDataValue(v: DFDLTime): DataValueTime = new DataValue(v)
+  @inline implicit def toDataValue(v: Array[Byte]): DataValueByteArray = new DataValue(v)
+  @inline implicit def toDataValue(v: JBoolean): DataValueBool = new DataValue(v)
+  @inline implicit def toDataValue(v: JNumber): DataValueNumber = new DataValue(v)
+  @inline implicit def toDataValue(v: JLong): DataValueLong = new DataValue(v)
+  @inline implicit def toDataValue(v: JDouble): DataValueDouble = new DataValue(v)
+  @inline implicit def toDataValue(v: JBigInt): DataValueBigInt = new DataValue(v)
+  @inline implicit def toDataValue(v: JString): DataValueString = new DataValue(v)
+  @inline implicit def toDataValue(v: URI): DataValueURI = new DataValue(v)
+  @inline implicit def toDataValue(v: JFloat): DataValueFloat = new DataValue(v)
+  @inline implicit def toDataValue(v: JByte): DataValueByte = new DataValue(v)
+  @inline implicit def toDataValue(v: JInt): DataValueInt = new DataValue(v)
+  @inline implicit def toDataValue(v: JShort): DataValueShort = new DataValue(v)
+  @inline implicit def toDataValue(v: DINode): DataValueDINode = new DataValue(v)
```

Comment:
Wait what? We shouldn't need this. v: DINode  what could this possibly mean.

Reply:
One of the things I avoided doing was distinguishing the infoset dataValue from DPath values.

Since DINode are legal as DPath values, they are therefore a primitive type.

I suppose we could distinguish this in the same was we distinguish UseNilForDefault, but I am not convinced it is worth the complexity. In particular, we will need to explicitly downcast whenever we want to assign a DPath value to an infoset dataValue. Since this downcast would not actually verify the type involved is not a DINode, creating this distinction might just give us complication and a false sense of security.

We can get away with it with UseNilForDefault because we only have 1 place where we need the downcast.

Reply:
Well, DINodes are never the values of DINodes. We may pass them in DPath to do things like get fn:namespace or something, but they never can be returned by say, inputValueCalc, or outputValueCalc. They can never become the value of an InfosetElement.

That's what makes this confusing. 

I would put in a comment like "Reserved for future use if we use DataValues in DPath."

Reply:
This isn't a future use. This commit uses the DataValue type as part of DPath:

Note arg1 below is a DINode.

```
abstract class FNTwoArgsNodeAndValue(recipes: List[CompiledDPath])
  extends RecipeOpWithSubRecipes(recipes) {
  override def run(dstate: DState) {

    val recipe1 = recipes(0)
    val recipe2 = recipes(1)

    val savedNode = dstate.currentNode
    dstate.resetValue()
    recipe1.run(dstate)
    val arg1 = dstate.currentNode

    dstate.setCurrentNode(savedNode)
    recipe2.run(dstate)
    val arg2 = dstate.currentValue.getNonNullable

    dstate.setCurrentValue(computeValue(arg1, arg2, dstate))
  }

  def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive

  override def toXML = toXML(recipes.map { _.toXML })
}
```

There is certainly merit in having a seperate type (alias) for types that are usable in the infoset and types that are usable within DPath; I'm just not convinced it will actually be worth it given how our code is currently structured.

Reply:
Yeah. The case you cite is for the dfdl:contentLength and dfdl:valueLength functions, and I suppose fn:count would be similar. Nodes there are arguments. Never results, but the DPath engine has to pass nodes around by way of creating suspensions that eventually compute their values. Methods there might even be returning nodes. So I guess I'm pretty well convinced.